gef.py

```diff
@@ -536,47 +536,233 @@ def __init__(self, elf="", minimalist=False):
-                    "{}QQQ".format(endian), fd.read(24))
-            else:
-                # else arch 32bits
-                self.e_entry, self.e_phoff, self.e_shoff = struct.unpack(
-                    "{}III".format(endian), fd.read(12))
+        if isinstance(elf, str):
+            if not os.access(elf, os.R_OK):
+                err("'{0}' not found/readable".format(elf))
+                err("Failed to get file debug information, most of gef features will not work")
+                return
+            self.fd = open(elf, "rb")
+            self.addr = None
+            self.pos = 0
+        elif isinstance(elf, int):
+            self.fd = None
+            self.addr = elf
+            self.pos = 0
+        else:
+            raise
+
+        # off 0x0
+        self.e_magic, self.e_class, self.e_endianness, self.e_eiversion = struct.unpack(">IBBB", self.read(7))
+
+        # adjust endianness in bin reading
+        endian = "<" if self.e_endianness == Elf.LITTLE_ENDIAN else ">"
+
+        # off 0x7
+        self.e_osabi, self.e_abiversion = struct.unpack("{}BB".format(endian), self.read(2))
+
+        # off 0x9
+        self.e_pad = self.read(7)
+
+        # off 0x10
+        self.e_type, self.e_machine, self.e_version = struct.unpack("{}HHI".format(endian), self.read(8))
+
+        # off 0x18
+        if self.e_class == Elf.ELF_64_BITS:
+            # if arch 64bits
+            self.e_entry, self.e_phoff, self.e_shoff = struct.unpack("{}QQQ".format(endian), self.read(24))
+        else:
+            # else arch 32bits
+            self.e_entry, self.e_phoff, self.e_shoff = struct.unpack("{}III".format(endian), self.read(12))
+
+        self.e_flags, self.e_ehsize, self.e_phentsize, self.e_phnum = struct.unpack("{}IHHH".format(endian), self.read(10))
+        self.e_shentsize, self.e_shnum, self.e_shstrndx = struct.unpack("{}HHH".format(endian), self.read(6))
+
+        self.phdrs = []
+        for i in range(self.e_phnum):
+            self.phdrs.append(Phdr(self, self.e_phoff + self.e_phentsize * i))
+
+        self.shdrs = []
+        for i in range(self.e_shnum):
+            self.shdrs.append(Shdr(self, self.e_shoff + self.e_shentsize * i))
+
+        if self.fd is not None:
+            self.fd.close()
+            self.fd = None
 
-            self.e_flags, self.e_ehsize, self.e_phentsize, self.e_phnum = struct.unpack(
-                "{}HHHH".format(endian), fd.read(8))
-            self.e_shentsize, self.e_shnum, self.e_shstrndx = struct.unpack(
-                "{}HHH".format(endian), fd.read(6))
+        return
+
+    def read(self, size):
+        if self.fd is not None:
+            v = self.fd.read(size)
+        elif self.addr is not None:
+            v = read_memory(self.addr + self.pos, size)
+        else:
+            raise
+        self.pos += size
+        return v
+
+    def seek(self, off):
+        if self.fd is not None:
+            self.fd.seek(off, 0)
+        elif self.addr is not None:
```

Comment:
```suggestion
        elif self.addr:
```

Reply:
I also thought about this, when converting the code, but it's not exactly the same, isn't it?

If for some weird case, the ELF would be mapped at 0x0, this would not work out, or?

Reply:
Removed memory parsing for now. It's not used/needed for now, and I had no proper testing environments, so we can think about it, when we really implement a new command that will be using that